TimeSpan for indicators

Worthy7 opened the pull request:
Allow indicators to accept TimeSpans so users can fine tune their indicators. (5m etc)

#### Description
Added a `ResolutionTimeSpan` class which implicitly converts between `Resolution` and `TimeSpan`, then F&R'd everything in QCAlgo.Indicators to use this new class instead for complete backwards compatibility. (Except perhaps indicator names changing)

#### Related Issue
https://github.com/QuantConnect/Lean/issues/1529

#### Motivation and Context
It allows fine tuning of indicator resolution

#### Requires Documentation Change
It would require the documentation to be appended to say that "TimeSpan.FromX can also be used"

#### How Has This Been Tested?
It's not yet.

#### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Non-functional change (xml comments/documentation/etc)

#### Checklist:
<!--- The following is a checklist of items that MUST be completed before a PR is accepted -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [x] I have read the **CONTRIBUTING** [document](https://github.com/QuantConnect/Lean/blob/master/CONTRIBUTING.md).
- [ ] I have added tests to cover my changes. <!--- If not applicable, please explain why -->
- [x] All new and existing tests passed.
- [ ] My branch follows the naming convention `bug-<issue#>-<description` or `feature-<issue#>-<description>`

<!--- Template inspired by https://www.talater.com/open-source-templates/#/page/99 -->

Worthy7 commented:
@mchandschuh Thanks for letting me know about conversions! They are pretty cool!
@AlexCatarino What do you think to using this instead?

Some points:
- Indicator names will all use the timespan style from now on
- I can't think of a decent test to write for this
- Intellisense will show `ResolutionTimeSpan` as the input type which may confuse some people.

AlexCatarino commented on Algorithm/QCAlgorithm.Indicators.cs:
```diff
@@ -1268,44 +1269,9 @@ public WilderMovingAverage WWMA(Symbol symbol, int period, Resolution? resolutio
-            }
-
-            return string.Format("{0}({1}{2})", type, symbol.ToString(), res);
+            return string.Format("{0}({1}{2})", type, symbol.ToString(), resolution.ToString());
```
You should keep the previous format ("_tick", "_sec", ..., "_day"). 
If it is not one of those cases, then we could use `.ToString()`.
You can add an unit test for this method.

Worthy7 replied:
Hmm but it will never be in the Resolution format when it reaches this function. The ResolutionTimespan class only converts to TimeSpan.
So it will always see TimeSpan and it won't match any of the cases.

AlexCatarino replied:
Why not something like this?
```csharp
if (resolution == TimeSpan.Zero)
{
   return $"{type}{symbol}_tick";
}
else if (resolution == QuantConnect.Time.OneSecond)
{
    return $"{type}{symbol}_sec";
}
else if (resolution == QuantConnect.Time.OneMinute)
{
    return $"{type}{symbol}_min";
}
// other cases
else
{
    return $"{type}{symbol}{resolution}";
}
```

mchandschuh replied:
There's a function for this :)  Checkout `Extensions.ToHigherResolutionEquivalent(false)` -- might be able to use the output of that, and then determine how many 'units' of the resolution fit into the timespan, so if you have 5 minutes, then `ToHigherResolutionEquivalent(false)` will return Resolution.Minute, and then 5minutes/Resolution.Minute.ToTimeSpan() => 5.

Anyway, this is certainly not incredibly important here

Worthy7 replied:
I still like my solution the most because there is no simple way to take care of 1:30:00.
What should 1 hour and 30 mins return?

Worthy7 commented:
I don't have SPY data on my local machine... Or do you want me to do this just to show the feature?

I haven't ran any tests yet sorry.
I did implement this into my own branch and it works fantastically, indicators all reporting the correct values.
I can run the tests and report if there are issues

mchandschuh commented on Common/Util/ResolutionTimeSpan.cs:
```diff
@@ -0,0 +1,31 @@
+﻿using System;
```
Please add license banner

mchandschuh commented on Common/Util/ResolutionTimeSpan.cs:
```diff
@@ -0,0 +1,31 @@
+    /// <summary>
+    /// This class allows implicit conversions between Resolution and TimeSpan
+    /// </summary>
+    public class ResolutionTimeSpan
```
Please add unit tests for this class. Also included under test should be use cases involving python.

mchandschuh replied:
I think this type is probably better defined as a `struct` since it represents a value (timespan is `struct` and enums are `structs` as well).

Worthy7 replied:
No idea how to write tests including use cases involving python.

mchandschuh commented on Common/Util/ResolutionTimeSpan.cs:
```diff
@@ -0,0 +1,31 @@
+    public class ResolutionTimeSpan
+    {
+        public ResolutionTimeSpan(TimeSpan timeSpan) { val = timeSpan; }
+        public TimeSpan val;
```
Since this should probably be a struct, let's make this `private readonly TimeSpan _value;` -- following along with the project's code style.

mchandschuh commented on Common/Util/ResolutionTimeSpan.cs:
```diff
@@ -0,0 +1,31 @@
+    /// </summary>
+    public class ResolutionTimeSpan
+    {
+        public ResolutionTimeSpan(TimeSpan timeSpan) { val = timeSpan; }
```
Please add XML documentation and format this function according to the prevailing project's code style.

mchandschuh commented:
An after though, perhaps we should call this new class `Period` or `Interval` or something less amalgamated.

Worthy7 commented:
@mchandschuh pick a name pls

Worthy7 commented:
I've got no idea how to do python things in here, it is somewhat out my scope of knowledge sorry.
After turning it into a struct I had to make a lot of changes elsewhere and python.cs started complaining a lot and I've ended up with "Cannot convert to PyObject" errors which I don't understand since I didn't change anything special. So I'm lost here.

Worthy7 commented:
Made all other other changes.
Not sure what tests to write, please commit - I don't care about learning about python integration in depth and I've already fixed this locally myself long ago, so make this as easy as possible for me please

mchandschuh commented:
@Worthy7  -- I didn't forsee the python errors, don't hate me, but let's revert the struct change and we can get this merged in :)

Worthy7 commented:
@mchandschuh I think think it was something to do with my end, perhaps I didn't have python something loaded? I'm not sure, but I managed to make it a struct anyway and it works.
Do you still want me to revert the struct then? It seems to be working.

mchandschuh commented:
@Worthy7  -- if you can't confirm that python is working properly then I'll need to look into it closer when myself or someone from the team has some time.

Worthy7 commented:
OK, so in that case perhaps it's best I leave this here and wait for
someone to clone the branch just to check python. Sorry but I have never
used python in QC - I can imagine how to do it in the web editor but no
idea how to run it in VS. if you can give me some simple instructions to
test it I'll give it a shot

On Tue, 7 Aug 2018 at 23:54, Michael <notifications@github.com> wrote:

> @Worthy7 <https://github.com/Worthy7> -- if you can't confirm that python
> is working properly then I'll need to look into it closer when myself or
> someone from the team has some time.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/QuantConnect/Lean/pull/2345#issuecomment-411085346>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AEirH2hqbe_eXs-rDGuTEoEvewi1K-Plks5uOaoHgaJpZM4VtP6t>
> .
>

Worthy7 commented:
Is it just a simple case of: Install python, use build.exe, run `python algoname.py` ?

Worthy7 commented:
Hi, I added the same changes to candlestick patterns too, and I ran all the tests in the master branch then on my branch, the results were exactly the same except my new test passed. All the python tests failed obvious in both cases because I guess I don't have it set up properly (module not found errors)

Master:
![image](https://user-images.githubusercontent.com/4762399/43886426-a0d9ad7e-9bf6-11e8-8064-c0d5e1dcdf71.png)

This branch:
![image](https://user-images.githubusercontent.com/4762399/43886437-a7e3c0d2-9bf6-11e8-9d7d-bff9491d7e08.png)

mchandschuh commented:
Here's a readme for getting python to work: https://github.com/QuantConnect/Lean/blob/master/Algorithm.Python/readme.md

the TL;DR is install python 3.6 and properly (manually) update the binaries in the `./packages`  to use the `.win` version (renamed to `.dll`).

We just need to confirm that python can properly handle this implicit conversion and then I think this guy is good to go!

Stick with it! You're on the home stretch!

Worthy7 commented:
I knew this was going to get messy lol

```
C:\Users\Ian\Source\Repos\QuantLeanian>python setup.py

    Python support in Lean with pythonnet
    =====================================

    Prerequisites:
        - Python 3.6 x64 with pip
        - LEAN: https://github.com/QuantConnect/Lean#installation-instructions
        - Visual C++ for Python: https://www.microsoft.com/en-us/download/details.aspx?id=44266
        - git

    It will update conda, pip, wheel, setuptools, pandas packages.


    Install/updates required packages
    ---------------------------------

[ pip ] Successfully installed/updated: wheel, pandas, pip, setuptools

    Install/updates pythonnet
    -------------------------

  Failed building wheel for pythonnet
Command ""C:\Program Files (x86)\Microsoft Visual Studio\Shared\Python36_64\python.exe" -u -c "import setuptools, tokenize;__file__='C:\\Users\\Ian\\AppData\\Local\\Temp\\pip-req-build-h1jkvyu0\\setup.py';f=getattr(tokenize, 'open', open)(__file__);code=f.read().replace('\r\n', '\n');f.close();exec(compile(code, __file__, 'exec'))" install --record C:\Users\Ian\AppData\Local\Temp\pip-record-tqabi0jm\install-record.txt --single-version-externally-managed --compile" failed with error code 1 in C:\Users\Ian\AppData\Local\Temp\pip-req-build-h1jkvyu0\
Collecting git+https://github.com/QuantConnect/pythonnet
  Cloning https://github.com/QuantConnect/pythonnet to c:\\users\\ian\\appdata\\local\\temp\\pip-req-build-h1jkvyu0
Building wheels for collected packages: pythonnet
  Running setup.py bdist_wheel for pythonnet: started
  Running setup.py bdist_wheel for pythonnet: finished with status 'error'
  Complete output from command "C:\\Program Files (x86)\\Microsoft Visual Studio\\Shared\\Python36_64\\python.exe" -u -c "import setuptools, tokenize;__file__=\'C:\\\\Users\\\\Ian\\\\AppData\\\\Local\\\\Temp\\\\pip-req-build-h1jkvyu0\\\\setup.py\';f=getattr(tokenize, \'open\', open)(__file__);code=f.read().replace(\'\\r\\n\', \'\\n\');f.close();exec(compile(code, __file__, \'exec\'))" bdist_wheel -d C:\\Users\\Ian\\AppData\\Local\\Temp\\pip-wheel-62nd87cn --python-tag cp36:
  running bdist_wheel
  running build
  running build_ext
  https://www.nuget.org/api/v2/ \x82\xcc\x8dX\x90V\x82\xf0\x8am\x94F\x82\xb5\x82\xc4\x82\xa2\x82\xdc\x82\xb7\x81B
  NuGet.exe 4.1.0 \x82\xcd\x8e\xc0\x8ds\x92\x86\x82\xc5\x82\xb7\x81B
  NuGet.exe \x82\xf0 4.6.2 \x82\xc9\x8dX\x90V\x82\xb5\x82\xc4\x82\xa2\x82\xdc\x82\xb7\x81B
  \x90\xb3\x8f\xed\x82\xc9\x8dX\x90V\x82\xb3\x82\xea\x82\xdc\x82\xb5\x82\xbd\x81B
  MSBuild auto-detection: using msbuild version '15.7.180.61344' from 'C:\\Program Files (x86)\\Microsoft Visual Studio\\2017\\Community\\MSBuild\\15.0\\bin'.
  Restoring NuGet package NUnit.3.7.1.
  Restoring NuGet package NUnit.ConsoleRunner.3.7.0.
  Restoring NuGet package UnmanagedExports.1.2.7.
  Adding package 'NUnit.3.7.1' to folder 'C:\\Users\\Ian\\AppData\\Local\\Temp\\pip-req-build-h1jkvyu0\\packages'
  Adding package 'NUnit.ConsoleRunner.3.7.0' to folder 'C:\\Users\\Ian\\AppData\\Local\\Temp\\pip-req-build-h1jkvyu0\\packages'
  Adding package 'UnmanagedExports.1.2.7' to folder 'C:\\Users\\Ian\\AppData\\Local\\Temp\\pip-req-build-h1jkvyu0\\packages'
  Added package 'UnmanagedExports.1.2.7' to folder 'C:\\Users\\Ian\\AppData\\Local\\Temp\\pip-req-build-h1jkvyu0\\packages'
  Added package 'NUnit.ConsoleRunner.3.7.0' to folder 'C:\\Users\\Ian\\AppData\\Local\\Temp\\pip-req-build-h1jkvyu0\\packages'
  Added package 'NUnit.3.7.1' to folder 'C:\\Users\\Ian\\AppData\\Local\\Temp\\pip-req-build-h1jkvyu0\\packages'

  NuGet Config files used:
      C:\\Users\\Ian\\AppData\\Local\\Temp\\pip-req-build-h1jkvyu0\\NuGet.Config
      C:\\Users\\Ian\\AppData\\Roaming\\NuGet\\NuGet.Config
      C:\\Program Files (x86)\\NuGet\\Config\\Microsoft.VisualStudio.Offline.config

  Feeds used:
      C:\\Users\\Ian\\.nuget\\packages\\
      https://dotnet.myget.org/F/dotnet-core/api/v3/index.json
      https://api.nuget.org/v3/index.json
      C:\\Program Files (x86)\\Microsoft SDKs\\NuGetPackages\\

  Installed:
      3 package(s) to packages.config projects
  Microsoft (R) Manifest Tool version 6.3.9600.17336
  Copyright (c) Microsoft Corporation 2012.
  All rights reserved.
  .NET Framework \x8c\xfc\x82\xaf Microsoft (R) Build Engine \x83o\x81[\x83W\x83\x87\x83\x93 15.7.180.61344
  Copyright (C) Microsoft Corporation.All rights reserved.

  .NET Framework \x8c\xfc\x82\xaf Microsoft (R) Build Engine \x83o\x81[\x83W\x83\x87\x83\x93 15.7.180.61344
  Copyright (C) Microsoft Corporation.All rights reserved.

  delegatemanager.cs(204,13): warning CS0162: \x93\x9e\x92B\x82\xc5\x82\xab\x82\xc8\x82\xa2\x83R\x81[\x83h\x82\xaa\x8c\x9f\x8fo\x82\xb3\x82\xea\x82\xdc\x82\xb5\x82\xbd [C:\\Users\\Ian\\AppData\\Local\\Temp\\pip-req-build-h1jkvyu0\\src\\runtime\\Python.Runtime.csproj]
  pyobject.cs(51,13): warning CS0162: \x93\x9e\x92B\x82\xc5\x82\xab\x82\xc8\x82\xa2\x83R\x81[\x83h\x82\xaa\x8c\x9f\x8fo\x82\xb3\x82\xea\x82\xdc\x82\xb5\x82\xbd [C:\\Users\\Ian\\AppData\\Local\\Temp\\pip-req-build-h1jkvyu0\\src\\runtime\\Python.Runtime.csproj]
  pythonexception.cs(67,13): warning CS0162: \x93\x9e\x92B\x82\xc5\x82\xab\x82\xc8\x82\xa2\x83R\x81[\x83h\x82\xaa\x8c\x9f\x8fo\x82\xb3\x82\xea\x82\xdc\x82\xb5\x82\xbd [C:\\Users\\Ian\\AppData\\Local\\Temp\\pip-req-build-h1jkvyu0\\src\\runtime\\Python.Runtime.csproj]
  pyscope.cs(534,13): warning CS0162: \x93\x9e\x92B\x82\xc5\x82\xab\x82\xc8\x82\xa2\x83R\x81[\x83h\x82\xaa\x8c\x9f\x8fo\x82\xb3\x82\xea\x82\xdc\x82\xb5\x82\xbd [C:\\Users\\Ian\\AppData\\Local\\Temp\\pip-req-build-h1jkvyu0\\src\\runtime\\Python.Runtime.csproj]
    Python.Runtime -> C:\\Users\\Ian\\AppData\\Local\\Temp\\pip-req-build-h1jkvyu0\\src\\runtime\\bin\\Python.Runtime.dll
    Python.Test -> C:\\Users\\Ian\\AppData\\Local\\Temp\\pip-req-build-h1jkvyu0\\src\\testing\\bin\\Python.Test.dll
  dynamic.cs(121,28): warning CS0618: 'PythonEngine.RunString(string, IntPtr?, IntPtr?)' \x82\xcd\x8b\x8c\x8c`\x8e\xae\x82\xc5\x82\xb7 ('RunString is deprecated and will be removed. Use Exec/Eval/RunSimpleString instead.') [C:\\Users\\Ian\\AppData\\Local\\Temp\\pip-req-build-h1jkvyu0\\src\\embed_tests\\Python.EmbeddingTest.csproj]
    Python.EmbeddingTest -> C:\\Users\\Ian\\AppData\\Local\\Temp\\pip-req-build-h1jkvyu0\\src\\embed_tests\\bin\\Python.EmbeddingTest.dll
    Console -> C:\\Users\\Ian\\AppData\\Local\\Temp\\pip-req-build-h1jkvyu0\\src\\console\\bin\\nPython.exe
    clrmodule -> C:\\Users\\Ian\\AppData\\Local\\Temp\\pip-req-build-h1jkvyu0\\src\\clrmodule\\bin\\clrmodule.dll
  C:\\Users\\Ian\\AppData\\Local\\Temp\\pip-req-build-h1jkvyu0\\packages\\UnmanagedExports.1.2.7\\tools\\RGiesecke.DllExport.targets(58,3): error MSB4062: "RGiesecke.DllExport.MSBuild.DllExportAppDomainIsolatedTask" \x83^\x83X\x83N\x82\xf0\x83A\x83Z\x83\x93\x83u\x83\x8a C:\\Users\\Ian\\AppData\\Local\\Temp\\pip-req-build-h1jkvyu0\\packages\\UnmanagedExports.1.2.7\\tools\\RGiesecke.DllExport.MSBuild.dll \x82\xa9\x82\xe7\x93\xc7\x82\xdd\x8d\x9e\x82\xdf\x82\xdc\x82\xb9\x82\xf1\x82\xc5\x82\xb5\x82\xbd\x81B\x83t\x83@\x83C\x83\x8b\x82\xdc\x82\xbd\x82\xcd\x83A\x83Z\x83\x93\x83u\x83\x8a \'Microsoft.Build.Utilities, Version=2.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a\'\x81A\x82\xdc\x82\xbd\x82\xcd\x82\xbb\x82\xcc\x88\xcb\x91\xb6\x8a\xd6\x8cW\x82\xcc 1 \x82\xc2\x82\xaa\x93\xc7\x82\xdd\x8d\x9e\x82\xdf\x82\xdc\x82\xb9\x82\xf1\x82\xc5\x82\xb5\x82\xbd\x81B\x8ew\x92\xe8\x82\xb3\x82\xea\x82\xbd\x83t\x83@\x83C\x83\x8b\x82\xaa\x8c\xa9\x82\xc2\x82\xa9\x82\xe8\x82\xdc\x82\xb9\x82\xf1\x81B <UsingTask> \x90\xe9\x8c\xbe\x82\xaa\x90\xb3\x82\xb5\x82\xa2\x82\xb1\x82\xc6\x81A\x83A\x83Z\x83\x93\x83u\x83\x8a\x82\xc6\x82\xbb\x82\xcc\x88\xcb\x91\xb6\x8a\xd6\x8cW\x82\xaa\x8eg\x97p\x89\xc2\x94\\\x82\xc5\x82\xa0\x82\xe9\x82\xb1\x82\xc6\x81A\x82\xa8\x82\xe6\x82\xd1 Microsoft.Build.Framework.ITask \x82\xf0\x8e\xc0\x91\x95\x82\xb7\x82\xe9\x83p\x83u\x83\x8a\x83b\x83N \x83N\x83\x89\x83X\x82\xaa\x83^\x83X\x83N\x82\xc9\x8a\xdc\x82\xdc\x82\xea\x82\xc4\x82\xa2\x82\xe9\x82\xb1\x82\xc6\x82\xf0\x8am\x94F\x82\xb5\x82\xc4\x82\xad\x82\xbe\x82\xb3\x82\xa2\x81B [C:\\Users\\Ian\\AppData\\Local\\Temp\\pip-req-build-h1jkvyu0\\src\\clrmodule\\clrmodule.csproj]
  Traceback (most recent call last):
    File "<string>", line 1, in <module>
    File "C:\\Users\\Ian\\AppData\\Local\\Temp\\pip-req-build-h1jkvyu0\\setup.py", line 527, in <module>
      zip_safe=False,
    File "C:\\Program Files (x86)\\Microsoft Visual Studio\\Shared\\Python36_64\\lib\\site-packages\\setuptools\\__init__.py", line 131, in setup
      return distutils.core.setup(**attrs)
    File "C:\\Program Files (x86)\\Microsoft Visual Studio\\Shared\\Python36_64\\lib\\distutils\\core.py", line 148, in setup
      dist.run_commands()
    File "C:\\Program Files (x86)\\Microsoft Visual Studio\\Shared\\Python36_64\\lib\\distutils\\dist.py", line 955, in run_commands
      self.run_command(cmd)
    File "C:\\Program Files (x86)\\Microsoft Visual Studio\\Shared\\Python36_64\\lib\\distutils\\dist.py", line 974, in run_command
      cmd_obj.run()
    File "C:\\Users\\Ian\\AppData\\Local\\Temp\\pip-req-build-h1jkvyu0\\setup.py", line 475, in run
      return bdist_wheel.bdist_wheel.run(self)
    File "C:\\Program Files (x86)\\Microsoft Visual Studio\\Shared\\Python36_64\\lib\\site-packages\\wheel\\bdist_wheel.py", line 202, in run
      self.run_command('build')
    File "C:\\Program Files (x86)\\Microsoft Visual Studio\\Shared\\Python36_64\\lib\\distutils\\cmd.py", line 313, in run_command
      self.distribution.run_command(command)
    File "C:\\Program Files (x86)\\Microsoft Visual Studio\\Shared\\Python36_64\\lib\\distutils\\dist.py", line 974, in run_command
      cmd_obj.run()
    File "C:\\Program Files (x86)\\Microsoft Visual Studio\\Shared\\Python36_64\\lib\\distutils\\command\\build.py", line 135, in run
      self.run_command(cmd_name)
    File "C:\\Program Files (x86)\\Microsoft Visual Studio\\Shared\\Python36_64\\lib\\distutils\\cmd.py", line 313, in run_command
      self.distribution.run_command(command)
    File "C:\\Program Files (x86)\\Microsoft Visual Studio\\Shared\\Python36_64\\lib\\distutils\\dist.py", line 974, in run_command
      cmd_obj.run()
    File "C:\\Program Files (x86)\\Microsoft Visual Studio\\Shared\\Python36_64\\lib\\distutils\\command\\build_ext.py", line 339, in run
      self.build_extensions()
    File "C:\\Program Files (x86)\\Microsoft Visual Studio\\Shared\\Python36_64\\lib\\distutils\\command\\build_ext.py", line 448, in build_extensions
      self._build_extensions_serial()
    File "C:\\Program Files (x86)\\Microsoft Visual Studio\\Shared\\Python36_64\\lib\\distutils\\command\\build_ext.py", line 473, in _build_extensions_serial
      self.build_extension(ext)
    File "C:\\Users\\Ian\\AppData\\Local\\Temp\\pip-req-build-h1jkvyu0\\setup.py", line 265, in build_extension
      subprocess.check_call(" ".join(cmd + ["/t:Build"]), shell=use_shell)
    File "C:\\Program Files (x86)\\Microsoft Visual Studio\\Shared\\Python36_64\\lib\\subprocess.py", line 291, in check_call
      raise CalledProcessError(retcode, cmd)
  subprocess.CalledProcessError: Command \'"C:\\Program Files (x86)\\Microsoft Visual Studio\\2017\\Community\\MSBuild\\15.0\\Bin\\MSBuild.exe" pythonnet.sln /p:Configuration=ReleaseWin /p:Platform=x64 /p:DefineConstants="PYTHON36%3BPYTHON3%3BUCS2" /p:PythonBuildDir="C:\\Users\\Ian\\AppData\\Local\\Temp\\pip-req-build-h1jkvyu0\\build\\lib.win-amd64-3.6" /p:PythonInteropFile="interop36.cs" /verbosity:minimal /p:PythonManifest="C:\\Users\\Ian\\AppData\\Local\\Temp\\pip-req-build-h1jkvyu0\\build\\lib.win-amd64-3.6\\app.manifest" /t:Build\' returned non-zero exit status 1.

  ----------------------------------------
  Running setup.py clean for pythonnet
Failed to build pythonnet
Installing collected packages: pythonnet
  Running setup.py install for pythonnet: started
    Running setup.py install for pythonnet: finished with status 'error'
    Complete output from command "C:\\Program Files (x86)\\Microsoft Visual Studio\\Shared\\Python36_64\\python.exe" -u -c "import setuptools, tokenize;__file__=\'C:\\\\Users\\\\Ian\\\\AppData\\\\Local\\\\Temp\\\\pip-req-build-h1jkvyu0\\\\setup.py\';f=getattr(tokenize, \'open\', open)(__file__);code=f.read().replace(\'\\r\\n\', \'\\n\');f.close();exec(compile(code, __file__, \'exec\'))" install --record C:\\Users\\Ian\\AppData\\Local\\Temp\\pip-record-tqabi0jm\\install-record.txt --single-version-externally-managed --compile:
    usage: -c [global_opts] cmd1 [cmd1_opts] [cmd2 [cmd2_opts] ...]
       or: -c --help [cmd1 cmd2 ...]
       or: -c --help-commands
       or: -c cmd --help

    error: option --single-version-externally-managed not recognized

    ----------------------------------------
```

mchandschuh commented:
@AlexCatarino  --  any comment on @Worthy7's pythonnet failure here? 

When I upgrade python I usually just do a rebuild, copy the .win to lib folder, renaming to .dll in the ./package/quantconnect.pythonnet.<version> and then rebuild again and it works (assuming you have python 3.6 installed of course)

AlexCatarino commented:
@Worthy7 , please checkout the updates we made in [setup.py](https://github.com/QuantConnect/Lean/blob/master/setup.py) and python [readme](https://github.com/QuantConnect/Lean/tree/master/Algorithm.Python#windows).
I suspect that you don't have .NET Framework 3.5 installed.

Worthy7 commented:
@mchandschuh  Can you please explain explicitly what you mean by renaming binaries. From what folder do I need to copy what files to where with what new names?
Thanks

Worthy7 commented:
I've gotten rid of the error, the setup.py works but 
`Message: System.DllNotFoundException : DLL 'python3.6m' を読み込めません:指定されたモジュールが見つかりません。 (HRESULT からの例外:0x8007007E)` when trying to run tests.
It says it `can't read the python3.6m DLL, Can't find the module` 

So I think I just need to copy some files... @AlexCatarino perhaps we should add this to the readme windows instructions

mchandschuh commented:
A step many people forget is to force rebuild all to ensure the new `Python.Runtime.dll` is copied to the required output directories.

Worthy7 commented:
@AlexCatarino  It's already there.
@mchandschuh I can see the dll in my bin/debug folder
I think I've spent enough time on this now guys, cant someone pull and test it please.

mchandschuh commented:
@AlexCatarino investigated supporting this in python and it didn't work as it is. The issue was that pythonnet immediately converts the `Resolution` to an integer value which then gets converted to a python `timedelta` using that value as the number of days. Alex mentioned he tried defining other implicit conversions to no avail. Alex will comment with further details about the attempts made for posterity.

Worthy7 commented:
Ok, thanks guys. Sucks that it doesn't "just work" but I appreciate you
taking the time to test it.

I suppose one solution is reverting to having multiple shortcut methods,
one for resolution, one for timespan. I think I still have that branch
somewhere, or I can throw it together again.


On Thu, 23 Aug 2018, 00:24 Alexandre Catarino, <notifications@github.com>
wrote:

> As @mchandschuh <https://github.com/mchandschuh> stated, I pulled @Worthy7
> <https://github.com/Worthy7> 's branch
> <https://github.com/Worthy7/Lean/tree/ResolutionTimeSpan> and ran the
> regression tests for python algorithms without success.
> My first attempt was to use int instead of TimeSpan since Resolution is
> an integer at pythonnet's "eyes". We would consider that values below 5
> represent a Resolution and other would represent microseconds. Since one
> day in microseconds are too big for int, I have used long instead. In
> this case, pythonnet wasn't able to find the implicit conversion method.
> My second attempt was to use double since it is a "direct" conversion.
> The same assumption was used to convert into a Resolution. In this case,
> pythonnet couldn't solve the implicit method overload considering an
> ambiguity.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/QuantConnect/Lean/pull/2345#issuecomment-415071915>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AEirH0Ux_Zuq26Qhg6xfWR5YHL5oj0IKks5uTXeagaJpZM4VtP6t>
> .
>

mchandschuh commented:
@Worthy7  --  the concern with the multiple methods is one of maintenance. Duplications of API methods are a pain to maintain and keep in sync with each other. I'm hopeful that we'll find a good solution to this, but I think the best solution involves making our version of pythonnet away of this special type so it can handle it properly which will involve it's own analysis and engineering effort.

@jaredbroad  -- what are your thoughts on having duplicate methods for indicator helpers, one for `Resolution` and one for `TimeSpan` ?  

If the duplicate methods is a non-starter, than we can close this PR and come back to it when pythonnet has improved implicit operator handling.

worthy7 commented:
btw, 4 years later and I am reviewing some of my own code - I _still_ think this PR is a good one. I find cases where I want my ATR/SM/RSI to give an update only on the 5 minute or 15 minute or 4 hour mark. That's the basis for this.
